src/main/scala/datalog/execution/PrecedenceGraph.scala

```diff
@@ -1,109 +1,132 @@
 package datalog.execution
 
 import datalog.dsl.Atom
-import datalog.tools.Debug.debug
 import datalog.storage.NS
+import datalog.tools.Debug.debug
 
 import scala.collection.mutable
 
-class Node(r: Int) (using ns: NS) {
-  var recursive = false
+private class Node(r: Int)(using ns: NS) {
   val rId: Int = r
   var idx: Int = -1
   var lowLink: Int = -1
   var edges: mutable.Set[Node] = mutable.Set[Node]()
   var onStack: Boolean = false
-  override def toString() =
+
+  def recursive: Boolean = edges.contains(this)
+
+  override def toString(): String =
     "{" + ns(rId) + ": " + "recursive=" + recursive
-//      " idx=" + idx + " lowLink=" + lowLink + " onstack=" + onStack +
-//      " edges=" + edges.map(e => ns(e.rId)).mkString("[", ", ", "]")
+      //      " idx=" + idx + " lowLink=" + lowLink + " onstack=" + onStack +
+      //      " edges=" + edges.map(e => ns(e.rId)).mkString("[", ", ", "]")
       + "}"
 }
 
 class PrecedenceGraph(using ns: NS /* for debugging */) {
-  val nodes: mutable.Map[Int, Node] = mutable.Map[Int, Node]()
-  val sorted: mutable.Queue[mutable.Set[Int]] = mutable.Queue[mutable.Set[Int]]()
-  val idbs: mutable.Set[Int] = mutable.Set[Int]()
+  private val adjacencyList = mutable.Map[Int, mutable.Set[Int]]()
```

Comment:
You may want to split the adjacency graph into two, one for positive and one for negative edges (maybe this is coming in a later PR that has to do specifically with negation and not just stratification, in which case you can ignore this comment). In order to check if a datalog program is stratisfiable, you would just check if there is a directed cycle of negative edges, which I think would make the most sense to be implemented here

Reply:
I'm a bit unsure about having the `PrecedenceGraph` handle the detection of non-stratifiable programs with negation. To me, it sounds like the `PrecedenceGraph` should only be responsible for computing the SCC of the relations, and the caller (probably the `ExecutionEngine` implementation) should then check that the suggested partial order of the strongly connected components actually satisfies the requirements for stratification (a simple `scan` keeping track of which relations have been computed in each step would be sufficient in that case).

That being said, I'm definitely planning to implement it in a follow-up PR 🙂 !